mypy/plugins/dataclasses.py

```diff
@@ -0,0 +1,291 @@
+from collections import OrderedDict
+from typing import Dict, List, Optional, Set, Tuple, cast
+
+from mypy.nodes import (
+    ARG_OPT, ARG_POS, MDEF, Argument, AssignmentStmt, Block, CallExpr, Context,
+    Decorator, Expression, JsonDict, NameExpr, SymbolTableNode, TempNode,
+    TypeInfo, Var
+)
+from mypy.plugin import ClassDefContext
+from mypy.plugins.common import _add_method, _get_decorator_bool_argument
+from mypy.types import CallableType, NoneTyp, Type, TypeVarDef, TypeVarType
+from mypy.typevars import fill_typevars
+
+# The set of decorators that generate dataclasses.
+dataclass_makers = {
+    'dataclass',
+    'dataclasses.dataclass',
+}
+
+
+class DataclassAttribute:
+    def __init__(
+            self,
+            name: str,
+            is_in_init: bool,
+            has_default: bool,
+            line: int, column: int,
+    ) -> None:
+        self.name = name
+        self.is_in_init = is_in_init
+        self.has_default = has_default
+        self.line = line
+        self.column = column
+
+    def to_argument(self, info: TypeInfo) -> Argument:
+        return Argument(
+            variable=self.to_var(info),
+            type_annotation=info[self.name].type,
+            initializer=None,
+            kind=ARG_OPT if self.has_default else ARG_POS,
+        )
+
+    def to_var(self, info: TypeInfo) -> Var:
+        return Var(self.name, info[self.name].type)
+
+    def serialize(self) -> JsonDict:
+        return {
+            'name': self.name,
+            'is_in_init': self.is_in_init,
+            'has_default': self.has_default,
+            'line': self.line,
+            'column': self.column,
+        }
```

Comment:
> Am I missing something obvious?

No, you are right, this way it will not work. Here is another idea about `InitVar`s that I think should work. You can just store _names_ of `InitVar`s in a special filed in `.metadata`, then later you can find the types in the type of `__init__` method, since in mypy both arg types and and arg names are stored. This is a bit hacky, but should be robust and is much simpler than manually traversing types in `fixup` phase.